### PR TITLE
Implemented the RNN decoder from the Karpathy+Fei-Fei paper

### PR DESCRIPTION
**Pull request by parinitaedke**
Wrote the MyRNNCell and MyRNN classes from the Karpathy and Fei-Fei paper.

Modified the SentenceDecoder class as well to reflect the changes in the RNN code.

### CODE REVIEW

File: Models/SentenceDecoder.py [hunk 1]
```diff
@@ -59,127 +61,133 @@ def forward(self, encoder_outputs, captions):
...
             self.rnn = LSTMDecoder(vocab_size=vocab_size, hidden_size=hidden_size)
         elif choice == 'RNN':
-            # not sure what obs_dim should be
-            self.rnn = MyRNN(obs_dim=obs_dim, hidden_size=hidden_size, output_dim=vocab_size)
+            self.rnn = MyRNN(vocab_size=vocab_size, hidden_size=hidden_size, device=device)
     
     def forward(self, captions, encoder_outputs):
-        return self.rnn.forward(captions=captions, encoder_outputs=encoder_outputs)
-        
+        # TODO: clarification - what is encoder_ouputs?  can't we get the size from captions?
```
Comment (Ikochoy) [inline Models/SentenceDecoder.py]: yes,please change it in this pr as well

Comment (parinitaedke) [reply Models/SentenceDecoder.py]: Done. Pushed a new version of it.

File: Models/SentenceDecoder.py [hunk 2]
```diff
@@ -51,135 +53,138 @@ def __init__(self, vocab_size, hidden_size):
...
+        # Send to GPU. Send Tensors created in a model to the same device as input Tensors.
+        embeddings = embeddings.to(self.device)
+        output_arr = output_arr.float().to(self.device)
+        hidden_arr = hidden_arr.float().to(self.device)
+
+        # Initialize RNN hidden state: h_0 is the 0 vector
+        hidden = torch.zeros(batch_size, self.hidden_size, device=self.device)
 
         for i in range(seq_len):
+            # h_0 is the 0 vector
```
Comment (Ikochoy) [inline Models/SentenceDecoder.py]: are these comments useful?

File: Models/SentenceDecoder.py [hunk 3]
```diff
@@ -51,135 +53,138 @@ def __init__(self, vocab_size, hidden_size):
...
         Returns predicted captions and its probability.
         """
         # batch_size x seq_len x hidden_size
         embeddings = self.embedding(captions)
         embeddings = torch.cat((torch.zeros(embeddings.shape[0], 1, embeddings.shape[2]), embeddings), dim=1)
+
         # batch_size x seq_len x hidden_size
-        h_t = torch.zeros(encoder_outputs.shape[0], 1, self.hidden_size)
-        c_t = torch.zeros(encoder_outputs.shape[0], 1, self.hidden_size)
+        h_t = torch.zeros(captions[0], 1, self.hidden_size)
```
Comment (Ikochoy) [inline Models/SentenceDecoder.py]: it should be encoder_outputs.shape[0], basically the image features shape [0]

File: Models/SentenceDecoder.py (hunk 2 above)
Comment (parinitaedke) [reply Models/SentenceDecoder.py]: Ah I had them in to help understand the flow of things. Can remove if needed.

File: Models/SentenceDecoder.py (hunk 3 above)
Comment (Ikochoy) [reply Models/SentenceDecoder.py]: by this you are assuming captions and image features have the same shape which might not be true. and captions[0] is the first caption in the list. but i am not sure whether captions.shape[0] works here as well

File: Models/SentenceDecoder.py (hunk 2 above)
Comment (Ikochoy) [reply Models/SentenceDecoder.py]: yea maybe remove them if they are not going to be useful

File: Models/SentenceDecoder.py (hunk 3 above)
Comment (parinitaedke) [reply Models/SentenceDecoder.py]: But aren't we just trying to get the batch size here? I meant to write captions.shape[0]. Wouldn't the batch size be the same regardless?

File: Models/SentenceDecoder.py (hunk 2 above)
Comment (parinitaedke) [reply Models/SentenceDecoder.py]: Ok will do.

File: Models/SentenceDecoder.py (hunk 3 above)
Comment (Ikochoy) [reply Models/SentenceDecoder.py]: not sure about this, and i think we need the image features? i don't think we can just use the shape of the image features, so i thought about it just now, i don't think we can remove encoder outputs. I think you will also need the encoder outputs for your rnn?

**Review by parinitaedke — COMMENTED** [review]
Looks good @Ikochoy

**Pull request merged**